File: mkdraft.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from datetime import datetime, timedelta
import json
import math
import os
import sys

from jinja2 import Environment, FileSystemLoader

import config
from utils import iso_period, jp_datespan

WORKSPACE = os.path.expanduser(config.WORKSPACE)

# round time to the nearest 5 minutes
def round_time(t): # datetime object
    s = 60 * (t.minute % 5) + t.second # seconds from the last 5-minute
    if s > 150 or (s == 150 and t.minute % 2 == 1): # round up
        t += timedelta(minutes=5)
    t -= timedelta(seconds=s)
    return t

# generate timeline
def gen_timeline(points): 
    ret = ""
    for i, p in enumerate(points):
        if i > 0:
            ret += ' …'
        ret += p['name']
        if 'ele' in p:
            ret += f"({p['ele']})"
        s = round_time(datetime.strptime(p['timespan'][0], '%Y-%m-%dT%H:%M:%S'))
        e = round_time(datetime.strptime(p['timespan'][1], '%Y-%m-%dT%H:%M:%S'))
        t = s if i > 0 else e
        ret += f" {t.hour}:{t.minute:02d}"
        if i > 0 and i < len(points) - 1:
            if e - s >= timedelta(minutes=10):
                ret += f"〜{e.hour}:{e.minute:02d}"
    return ret

# WGPS84 to Web Mercator projection
def trans(lon, lat):
    wpx = (lon + 180) / 360
    s = math.sin(math.radians(lat))
    wpy = 0.5 - (1 / (4 * math.pi)) * math.log((1 + s) / (1 - s))
    return wpx, wpy

# calculate center and zoom level for given bbox
def center(min_lon, min_lat, max_lon, max_lat): # bbox
    lon = (min_lon + max_lon) / 2
    lat = (min_lat + max_lat) / 2
    min_wp = trans(min_lon, max_lat) # northwest center
    max_wp = trans(max_lon, min_lat) # southeast center
    wx = 256 * (max_wp[0] - min_wp[0])
    wy = 256 * (max_wp[1] - min_wp[1])
    # NOTE: 580x400 is the size of the map frame
    xw = 580 / wx
    yw = 400 / wy
    w = min(xw, yw)
    zoom = int(math.log(w) / math.log(2))
    if zoom > 16:
        zoom = 16
    return round(lat, 6), round(lon, 6), zoom

# generate routemap URL
def gen_routemap(routemap):
    file = f"{WORKSPACE}/{cid}/{routemap}"
    with open(file, 'r', encoding='utf-8') as f:
        data = json.load(f)
    if not 'bbox' in data:
        print(f"Error: {file} does not have bbox.", file=sys.stderr)
        sys.exit(1)
    bbox = data['bbox']
    lat, lon, zoom = center(*bbox)
    return f"routemap.html?lat={lat}&amp;lon={lon}&amp;zoom={zoom}&amp;url={cid}/{routemap}"

# generate sequence of photo pairs
def gen_photo(photos):
    ret = []
    for i in range(0,len(photos),2):
        p0 = photos[i]
        p1 = photos[i+1]
        w0, h0 = (270, 180) if p0['width'] > p0['height'] else (180, 270)
        w1, h1 = (270, 180) if p1['width'] > p1['height'] else (180, 270)
        ret.append({
            'base0': p0['hash'], 'w0': w0, 'h0': h0, 'cap0': p0['caption'],
            'base1': p1['hash'], 'w1': w1, 'h1': h1, 'cap1': p1['caption']
        })
    return ret

# generate sequence of sections
def gen_section(sections):
    ret = []
    for s in sections:
        ret.append({
            'title': s['title'],
            'date': s['date'], # %Y-%m-%d
            'timeline': gen_timeline(s['timeline']),
            'routemap': gen_routemap(s['routemap']),
            'photo': gen_photo(s['photo'])
        })
    return ret

# command line arguments
if len(sys.argv) != 2:
    print(f"Usage: {sys.argv[0]} <cid>", file=sys.stderr)
    sys.exit(1)
cid = sys.argv[1] # Content ID
description = '⚠️ This article is a draft'

# load resource json
with open(f"{WORKSPACE}/{cid}.json", 'r', encoding='utf-8') as f:
    resource = json.load(f)

# set template context
s = datetime.strptime(resource['date']['start'], '%Y-%m-%d')
e = datetime.strptime(resource['date']['end'], '%Y-%m-%d')
now = datetime.now()

context = {
    'description': description,
    'title': resource['title'],
    'cid' : resource['cid'],
    'cover': resource['cover']['hash'],
    'period': iso_period(s, e),
    'pubdate': now.strftime('%Y-%m-%d'),
    'date': resource['date'], # {'start': '%Y-%m-%d', 'end': '%Y-%m-%d'}
    'datejp': (lambda args: {'start': args[0], 'end': args[1]})(jp_datespan(s, e)),
    'section': gen_section(resource['section']),
    'lm_year': now.year,
    'year': s.year
}

# Jinja2 template rendering
env = Environment(loader=FileSystemLoader('template'), trim_blocks=True)
template = env.get_template('draft.html')
with open(f"{WORKSPACE}/{cid}.html", "w", encoding='utf-8') as f:
    f.write(template.render(context))
    f.write('\n')
# __END__

File: attic/mkdraft.pl
#!/usr/bin/env -S perl -CA
# mkdraft.pl - Create draft HTML file

use strict;
use warnings;
use utf8;
use open qw(:utf8 :std);
use feature qw(say);

use Config::Tiny;
use JSON;
use Math::Trig qw(pi deg2rad);
use Text::Xslate;
use Time::Piece;

# Set days of the week in Japanese
Time::Piece::day_list(qw(日 月 火 水 木 金 土));

# Load configuration file
my $cf = Config::Tiny->read('config.ini');
my $workspace = $cf->{_}->{workspace};
my $material_gpx = $cf->{material}->{gpx};
my $material_img = $cf->{material}->{img};

# Parse command line argument
if ($#ARGV < 0) {
  my $script = basename($0);
  die "Usage: $script <CID>\n";
}
my $cid = $ARGV[0]; # content ID

# Load resource JSON file
my $file = "$workspace/$cid.json";
open(my $in, '<:raw', $file) or die "Can't open $file: $!\n";
my $text = do { local $/; <$in> };
close($in);
my $resource = decode_json($text);

# Convert start and end date to ISO8601 period
sub datespan {
  my $date = shift;
  my ($start, $end) = ($date->{start}, $date->{end});
  my ($ys, $ms, $ds) = split(/-/, $start);
  my ($ye, $me, $de) = split(/-/, $end);
  if ($ys ne $ye) {
    return "$start/$end";
  }
  if ($ms ne $me) {
    return "$start/$me-$de";
  }
  if ($ds ne $de) {
    return "$start/$de";
  }
  return $start;
}

# Convert start and end date object to Japanese period
sub datejp {
  my ($s, $e) = @_; # Time::Piece object
  my ($start, $end);
  $start = sprintf '%d年%d月%d日（%s）', $s->year, $s->mon, $s->mday, $s->wdayname;
  if ($s->year ne $e->year) {
    $end = sprintf '%d年%d月%d日（%s）', $e->year, $e->mon, $e->mday, $e->wdayname;
  } elsif ($s->mon ne $e->mon) {
    $end = sprintf '%d月%d日（%s）', $e->mon, $e->mday, $e->wdayname;
  } elsif ($s->mday ne $e->mday) {
    $end = sprintf '%d日（%s）', $e->mday, $e->wdayname;
  } else {
    $end = '';
  }
  return { start => $start, end => $end };
}

# Round time to nearest 5-minute interval
sub round_time {
  my $t = shift; # Time::Piece object
  my $s = 60 * ($t->minute % 5) + $t->sec;
  if ($s > 150 || $s == 150 && $t->minute % 2 == 1) {
    $t += 300;
  }
  $t -= $s;
  return $t;
}

# Create timeline
sub gen_timeline {
  my $points = shift;
  my $ret = '';
  for (my $i = 0; $i <= $#{$points}; $i++) {
    my $p = $points->[$i];
    if ($i > 0) {
      $ret .= ' …';
    }
    $ret .= $p->{name};
    if (exists $p->{ele}) {
      $ret .= "($p->{ele})";
    }
    my $s = round_time(Time::Piece->strptime($p->{timespan}[0], '%FT%T'));
    my $e = round_time(Time::Piece->strptime($p->{timespan}[1], '%FT%T'));
    my $t = $i > 0 ? $s : $e;
    $ret .= sprintf ' %d:%02d', $t->hour, $t->minute;
    if ($i > 0 && $i < $#{$points}) {
      my $diff = $e - $s;
      if ($diff->minutes >= 10) {
        $ret .= sprintf '〜%d:%02d', $e->hour, $e->minute;
      }
    }
  }
  return $ret;
}

# Convert latitude and longitude to Web Mercator coordinates
sub trans {
  my ($lon, $lat) = @_;
  my $wpx = ($lon + 180) / 360;
  my $s = sin(deg2rad($lat));
  my $wpy = 0.5 - (1 / (4 * pi)) * log((1 + $s) / (1 - $s));
  return ($wpx, $wpy);
}

# Calculate center coordinates and zoom level
sub center {
  my ($min_lon, $min_lat, $max_lon, $max_lat) = @_; # bounding box
# center coordinates
  my $lon = sprintf("%.6f", 0.5 * ($max_lon + $min_lon));
  my $lat = sprintf("%.6f", 0.5 * ($max_lat + $min_lat));
# zoom level
  my @min_wp = trans($min_lon, $max_lat); # north-west corner
  my @max_wp = trans($max_lon, $min_lat); # south-east corner
  my $wx = 256 * ($max_wp[0] - $min_wp[0]);
  my $wy = 256 * ($max_wp[1] - $min_wp[1]);
# NOTE: Map size is fixed at 580 x 400 pixels
  my $xw = 580 / $wx;
  my $yw = 400 / $wy;
  my $w = $xw < $yw ? $xw : $yw;
  my $zoom = int(log($w) / log(2));
  $zoom = 16 if $zoom > 16;
  return ($lat, $lon, $zoom);
}

# Create URL for routemap
sub gen_routemap {
  my $routemap = shift;
  my $file = "$workspace/$cid/$routemap";
  open(my $in, '<:raw', $file) or die "Can't open $file: $!\n";
  my $text = do { local $/; <$in> };
  close($in);
  my $json = decode_json($text);
  die "missing 'bbox' in $file\n"  unless exists $json->{bbox};
  my ($lat, $lon, $zoom) = center(@{$json->{bbox}});
  return "routemap.html?lat=$lat&lon=$lon&zoom=$zoom&url=$cid/$routemap";
}

# Create photo information
sub gen_photo {
  my $photos = shift;
  my $ret = [];
  for (my $i = 0; $i <= $#{$photos}; $i += 2) {
    my $p0 = $photos->[$i];
    my $p1 = $photos->[$i + 1];
    my ($w0, $h0) = $p0->{width} > $p0->{height} ? (270, 180) : (180, 270);
    my ($w1, $h1) = $p1->{width} > $p1->{height} ? (270, 180) : (180, 270);
    push @$ret, {
      base0 => $p0->{hash}, w0 => $w0, h0 => $h0, cap0 => $p0->{caption},
      base1 => $p1->{hash}, w1 => $w1, h1 => $h1, cap1 => $p1->{caption}
    };
  }
  return $ret;
}

# Create section
sub gen_section {
  my $sect = shift;
  my $ret = [];
  foreach my $s (@{$sect}) {
    push @$ret, {
      title => $s->{title},
      date => $s->{date},
      timeline => gen_timeline($s->{timeline}),
      routemap => gen_routemap($s->{routemap}),
      photo => gen_photo($s->{photo})
    };
  }
  return $ret;
}

# Set template variables
my $s = Time::Piece->strptime($resource->{date}->{start}, '%F');
my $e = Time::Piece->strptime($resource->{date}->{end}, '%F');
my $now = localtime;
my $template = $cf->{template};
my $vars = {
  description => 'なんたら、かんたら。', # This and that.
  title => $resource->{title},
  cid => $resource->{cid},
  cover => $resource->{cover}->{hash},
  datespan => datespan($resource->{date}),
  pubdate => $now->strftime('%F'),
  date => $resource->{date},
  datejp => datejp($s, $e),
  section => gen_section($resource->{section}),
  lm_year => $now->year,
  year => $s->year,
};

# Translate template
my $tx = Text::Xslate->new(syntax => 'TTerse', verbose => 2);
my $html = "$workspace/$cid.html";
say $html;
open(my $out, '>', $html) or die "Can't open $html: $!\n";
print $out $tx->render('tmpl/tozan.html', $vars);
close($out);
__END__
